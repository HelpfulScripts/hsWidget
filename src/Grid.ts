/**
 * # Grid widget
 * exposes CSS Grid Layout. 
 * 
 * ### Profile
 * invoked as 
 * - `m(Grid, <GridAttrs>, <content>);`
 * 
 * See {@link Grid.GridAttrs GridAttrs}
 * 
 * 
 * ### Example
 * <example height=250px><file name='script.js'>ximport='Grid.x.js'</file>
 * <file name='style.css'>
 * .gridExample { }
 * .colGrid { background-color:#eef; width:100%; height:100%; }
 * .rowGrid { background-color:#efe; padding:5px; width:100%; height:100%; }
 * .rowGrid .hs_grid_cell:nth-child(even) {background-color:#eee;}
 * .rowGrid .hs_grid_cell:nth-child(odd)  {background-color:#fff;}
 * .rowGrid div, .colGrid div { text-align:center; vertical-align:middle}
 * </file>
 * </example>
 * 
 */

/** */
import m                from "mithril";
import { Vnode }        from './Widget';
import { ViewResult }   from './Widget';
import { Widget }       from './Widget';
import { WidgetAttrs }  from './Widget';

export interface GridAttrs extends WidgetAttrs {
    /** grid row template. will be passed through as is to grid-template-rows command of the CSS Grid styling */
    rows?: string;
    /** grid row template. will be passed through as is to grid-template-columns command of the CSS Grid styling */
    columns?: string;
    /** grid-gap; the string will be passed through as is to the CSS Grid styling */
    gap?: string;
}


// type RC = 'row'|'col';

// const opp = {col:'row', row:'col'};

/**
 * # Grid
 */
export class Grid extends Widget {
    rc:string;
    gridFormat(a:GridAttrs) {
        // const template = {row:'rows', col:'columns'};
        // const isTmp = a.template!==undefined && a.template!=='';
        // const tmpRows = a.rows;
        // const tmpCols = a.columns || tmpRows? undefined : '';
        const cl = (a.rows===undefined && a.columns===undefined)? `hs_grid_auto_col` : (a.rows===undefined? 'hs_grid_col' : 'hs_grid_row');
        const st = (a.gap? `gap:${a.gap};`:'') +
                   (a.rows===undefined?   '' : `grid-template-rows:${a.rows};`) +
                   (a.columns===undefined?'' : `grid-template-columns:${a.columns};`);
        return [cl, st===''?undefined:st];
    }
    view(node: Vnode<GridAttrs, this>):ViewResult { 
        const a:GridAttrs = node.attrs;
        const [cl, st] = this.gridFormat(a);
        const childNodes = () => (<any[]>node.children).map((c,i) => typeof c==='string'? m(`.hs_grid_cell.child${i}`, c) : c);
        return m(`.hs_grid`, this.attrs(node.attrs, { class: cl, style:st}), childNodes());
    }
}

/**
 * # GridColumns 
 * Exposes the CSS Grid Layout to Typescript.
 * `m(Grid, <{@link Grid.GridAttrs `GridAttrs`}>, <content>);`
 */
// export class GridColumns extends Grid {
//     rc = 'col';
// }

/**
 * # GridRows 
 * Exposes the CSS Grid Layout to Typescript.
 * `m(Grid, <{@link Grid.GridAttrs `GridAttrs`}>, <content>);`
 */
// export class GridRows extends Grid {
//     rc = 'row';
// }
