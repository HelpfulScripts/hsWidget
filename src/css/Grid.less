
.hs_grid { 
    display:grid; 
    box-sizing: border-box;
    place-items: stretch;
    &.hs_grid_col { width:  100vw; grid-auto-flow: column; }
    &.hs_grid_row { height: 100vh; grid-auto-flow: row; }    
}

.hs_grid_auto_col {
    grid-auto-flow: column;
}
.hs_grid_auto_row {
    grid-auto-flow: row;
}