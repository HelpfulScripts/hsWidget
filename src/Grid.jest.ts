import m        from "mithril";
import { Grid } from "./Grid";

window = Object.assign(require('mithril/test-utils/domMock.js')(), require('mithril/test-utils/pushStateMock')());
const mq = require('mithril-query');


describe('Grid', () => {
    describe('columns', () => {
        const cells = ['1', '2', '3', '4', '5'];
        const out = mq(m(Grid, { 
            class: 'colGrid',
            gap:"5px 5px",
            columns:'50px auto 50px',
        }, [
            '<',
            m(Grid, {class:'rowGrid', rows:''}, cells),
            '>'
        ]));
        it ('should have DOM structure', () => {
            out.should.have('.hs_grid.colGrid');
            expect(out.rootNode.attrs.style).toBeDefined();
            expect(out.rootNode.attrs.style).toContain('grid-template-columns:50px auto 50px;');
            expect(out.rootNode.attrs.style).toContain('gap:5px 5px;');
            expect(out.rootNode).toMatchSnapshot();

            out.should.have('.hs_grid.colGrid .child0');
            out.should.have('.hs_grid.colGrid .rowGrid');
            out.should.have('.hs_grid.colGrid .child2');

            out.should.have('.rowGrid .child0');
            out.should.have('.rowGrid .child4');
            out.should.not.have('.rowGrid .child5');

        });
    });
    describe('rows', () => {
        const cells = ['1', '2', '3', '4', '5'];
        const out = mq(m(Grid, { 
            class: 'colGrid',
            gap:"5px 5px",
            rows:'50px auto 50px',
        }, [
            '<',
            m(Grid, {class:'rowGrid', rows:''}, cells),
            '>'
        ]));
        it ('should have DOM structure', () => {
            out.should.have('.hs_grid.colGrid');
            expect(out.rootNode.attrs.style).toBeDefined();
            expect(out.rootNode.attrs.style).toContain('grid-template-rows:50px auto 50px;');
            expect(out.rootNode.attrs.style).toContain('gap:5px 5px;');
            expect(out.rootNode).toMatchSnapshot();

            out.should.have('.hs_grid.colGrid .child0');
            out.should.have('.hs_grid.colGrid .rowGrid');
            out.should.have('.hs_grid.colGrid .child2');

            out.should.have('.rowGrid .child0');
            out.should.have('.rowGrid .child4');
            out.should.not.have('.rowGrid .child5');

        });
    });
    describe('nested', () => {
        const cells = ['1', '2', '3', '4', '5'];
        const out = mq(m(Grid, { 
            class: 'colGrid',
            style:'background-color:#ffe; padding:20px 0;',
            gap:"5px 5px",
            columns:'50px auto auto 50px', 
        }, [
            '<',
            m(Grid, {class:'rowGrid', rows:'' }, cells),
            m(Grid, {class:'rowGrid', rows:'' }, cells),
            '>'
        ]));
        it ('should have DOM structure', () => {
            out.should.have('.hs_grid.colGrid');
            expect(out.rootNode.attrs.style).toBeDefined();
            expect(out.rootNode.attrs.style).toContain('grid-template-columns:50px auto auto 50px;');
            expect(out.rootNode.attrs.style).toContain('gap:5px 5px;');
            expect(out.rootNode).toMatchSnapshot();

            out.should.have('.hs_grid.colGrid .child0');
            out.should.have('.hs_grid.colGrid .rowGrid');
            out.should.have('.hs_grid.colGrid .child2');

            out.should.have('.rowGrid .child0');
            out.should.have('.rowGrid .child4');
            out.should.not.have('.rowGrid .child5');

        });
    });
});
