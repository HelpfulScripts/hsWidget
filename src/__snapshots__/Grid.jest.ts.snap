// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid columns should have DOM structure 1`] = `
Object {
  "attrs": Object {
    "class": null,
    "className": "hs_grid colGrid hs_grid_col",
    "columns": "50px auto 50px",
    "gap": "5px 5px",
    "style": " gap:5px 5px;grid-template-columns:50px auto 50px;",
  },
  "children": Array [
    Object {
      "attrs": Object {
        "className": "hs_grid_cell child0",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "state": undefined,
      "tag": "div",
      "text": "<",
    },
    Object {
      "attrs": Object {
        "class": "rowGrid hs_grid_row",
        "rows": "",
        "style": " grid-template-rows:;",
      },
      "children": Array [
        "1",
        "2",
        "3",
        "4",
        "5",
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": Grid {},
      "key": undefined,
      "state": Grid {},
      "tag": [Function],
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "className": "hs_grid_cell child2",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "state": undefined,
      "tag": "div",
      "text": ">",
    },
  ],
  "dom": undefined,
  "domSize": undefined,
  "events": undefined,
  "instance": undefined,
  "key": undefined,
  "parent": undefined,
  "renderedChildren": Array [
    Object {
      "attrs": Object {
        "className": "hs_grid_cell child0",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "state": undefined,
      "tag": "div",
      "text": "<",
    },
    Object {
      "attrs": Object {
        "class": null,
        "className": "hs_grid rowGrid hs_grid_row",
        "rows": "",
        "style": " grid-template-rows:;",
      },
      "children": Array [
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child0",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "1",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child1",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "2",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child2",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "3",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child3",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "4",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child4",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "5",
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "renderedChildren": Array [
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child0",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "1",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child1",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "2",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child2",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "3",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child3",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "4",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child4",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "5",
        },
      ],
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "className": "hs_grid_cell child2",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "state": undefined,
      "tag": "div",
      "text": ">",
    },
  ],
  "state": undefined,
  "tag": "div",
  "text": undefined,
}
`;

exports[`Grid nested should have DOM structure 1`] = `
Object {
  "attrs": Object {
    "class": null,
    "className": "hs_grid colGrid hs_grid_col",
    "columns": "50px auto auto 50px",
    "gap": "5px 5px",
    "style": "background-color:#ffe; padding:20px 0; gap:5px 5px;grid-template-columns:50px auto auto 50px;",
  },
  "children": Array [
    Object {
      "attrs": Object {
        "className": "hs_grid_cell child0",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "state": undefined,
      "tag": "div",
      "text": "<",
    },
    Object {
      "attrs": Object {
        "class": "rowGrid hs_grid_row",
        "rows": "",
        "style": " grid-template-rows:;",
      },
      "children": Array [
        "1",
        "2",
        "3",
        "4",
        "5",
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": Grid {},
      "key": undefined,
      "state": Grid {},
      "tag": [Function],
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "class": "rowGrid hs_grid_row",
        "rows": "",
        "style": " grid-template-rows:;",
      },
      "children": Array [
        "1",
        "2",
        "3",
        "4",
        "5",
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": Grid {},
      "key": undefined,
      "state": Grid {},
      "tag": [Function],
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "className": "hs_grid_cell child3",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "state": undefined,
      "tag": "div",
      "text": ">",
    },
  ],
  "dom": undefined,
  "domSize": undefined,
  "events": undefined,
  "instance": undefined,
  "key": undefined,
  "parent": undefined,
  "renderedChildren": Array [
    Object {
      "attrs": Object {
        "className": "hs_grid_cell child0",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "state": undefined,
      "tag": "div",
      "text": "<",
    },
    Object {
      "attrs": Object {
        "class": null,
        "className": "hs_grid rowGrid hs_grid_row",
        "rows": "",
        "style": " grid-template-rows:;",
      },
      "children": Array [
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child0",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "1",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child1",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "2",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child2",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "3",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child3",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "4",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child4",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "5",
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "renderedChildren": Array [
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child0",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "1",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child1",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "2",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child2",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "3",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child3",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "4",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child4",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "5",
        },
      ],
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "class": null,
        "className": "hs_grid rowGrid hs_grid_row",
        "rows": "",
        "style": " grid-template-rows:;",
      },
      "children": Array [
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child0",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "1",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child1",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "2",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child2",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "3",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child3",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "4",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child4",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "5",
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "renderedChildren": Array [
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child0",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "1",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child1",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "2",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child2",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "3",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child3",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "4",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child4",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "5",
        },
      ],
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "className": "hs_grid_cell child3",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "state": undefined,
      "tag": "div",
      "text": ">",
    },
  ],
  "state": undefined,
  "tag": "div",
  "text": undefined,
}
`;

exports[`Grid rows should have DOM structure 1`] = `
Object {
  "attrs": Object {
    "class": null,
    "className": "hs_grid colGrid hs_grid_row",
    "gap": "5px 5px",
    "rows": "50px auto 50px",
    "style": " gap:5px 5px;grid-template-rows:50px auto 50px;",
  },
  "children": Array [
    Object {
      "attrs": Object {
        "className": "hs_grid_cell child0",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "state": undefined,
      "tag": "div",
      "text": "<",
    },
    Object {
      "attrs": Object {
        "class": "rowGrid hs_grid_row",
        "rows": "",
        "style": " grid-template-rows:;",
      },
      "children": Array [
        "1",
        "2",
        "3",
        "4",
        "5",
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "instance": Grid {},
      "key": undefined,
      "state": Grid {},
      "tag": [Function],
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "className": "hs_grid_cell child2",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "state": undefined,
      "tag": "div",
      "text": ">",
    },
  ],
  "dom": undefined,
  "domSize": undefined,
  "events": undefined,
  "instance": undefined,
  "key": undefined,
  "parent": undefined,
  "renderedChildren": Array [
    Object {
      "attrs": Object {
        "className": "hs_grid_cell child0",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "state": undefined,
      "tag": "div",
      "text": "<",
    },
    Object {
      "attrs": Object {
        "class": null,
        "className": "hs_grid rowGrid hs_grid_row",
        "rows": "",
        "style": " grid-template-rows:;",
      },
      "children": Array [
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child0",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "1",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child1",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "2",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child2",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "3",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child3",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "4",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child4",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "5",
        },
      ],
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "renderedChildren": Array [
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child0",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "1",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child1",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "2",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child2",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "3",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child3",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "4",
        },
        Object {
          "attrs": Object {
            "className": "hs_grid_cell child4",
          },
          "children": undefined,
          "dom": undefined,
          "domSize": undefined,
          "events": undefined,
          "inspect": [Function],
          "instance": undefined,
          "key": undefined,
          "parent": [Circular],
          "state": undefined,
          "tag": "div",
          "text": "5",
        },
      ],
      "state": undefined,
      "tag": "div",
      "text": undefined,
    },
    Object {
      "attrs": Object {
        "className": "hs_grid_cell child2",
      },
      "children": undefined,
      "dom": undefined,
      "domSize": undefined,
      "events": undefined,
      "inspect": [Function],
      "instance": undefined,
      "key": undefined,
      "parent": [Circular],
      "state": undefined,
      "tag": "div",
      "text": ">",
    },
  ],
  "state": undefined,
  "tag": "div",
  "text": undefined,
}
`;
